ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.cpp

```diff
@@ -10,37 +10,117 @@
 #include "ttnn/operations/data_movement/transpose/transpose.hpp"
 #include "ttnn/cpp/ttnn/operations/copy.hpp"
+#include "ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp"
+#include "ttnn/cpp/ttnn/operations/data_movement/slice/slice.hpp"
 
 
 namespace ttnn::operations::data_movement {
 
 namespace detail {
+uint32_t round_up(uint32_t value, uint32_t multiple) {
+    if (multiple == 0) {
+        return value;
+    } else if (value == 0) {
+        return multiple;
+    }
+    return (value % multiple == 0) ? value : ((value + multiple - 1) / multiple) * multiple;
+}
+
 inline Tensor transpose_(const Tensor &a, TransposeOpDim transpose_dim, const MemoryConfig& output_mem_config) {
     bool pad_c = false;
+    bool tiled_only = false;
     bool pad_n = false;
-
+    uint32_t ROW_MAJOR_STICK_WIDTH = 16; // this is a highly restrictive constraint on the RM transpose_wh kernel, and with all the other bugs/limitations we should rewrite it
+    uint32_t W = a.get_shape().with_tile_padding()[-1], H = a.get_shape().with_tile_padding()[-2];
     switch (transpose_dim) {
         case TransposeOpDim::HC:
-            pad_c = true;
+            tiled_only = a.get_layout() == Layout::TILE;
+            if ((!tiled_only) && ((W * a.element_size()) % ROW_MAJOR_STICK_WIDTH != 0)) { // for some reason this is a constraint on the RM transpose_hw kernel
+                tiled_only = true;
+            }
+            pad_c = tiled_only && a.get_shape().with_tile_padding()[1] % 32 != 0;
             break;
+        // bubble dim around to make it possible as these implementations don't have a kernel
         case TransposeOpDim::NH:
             return ttnn::permute((const ttnn::Tensor)a, std::vector<int64_t>({2, 1, 0, 3}), output_mem_config);
         case TransposeOpDim::NW:
             return ttnn::permute((const ttnn::Tensor)a, std::vector<int64_t>({3, 1, 2, 0}), output_mem_config);
         case TransposeOpDim::CW:
             return ttnn::permute((const ttnn::Tensor)a, std::vector<int64_t>({0, 3, 2, 1}), output_mem_config);
+        case TransposeOpDim::CN:
+            tiled_only = true; // CN only has a tiled implementation at the moment
+            break;
+        case TransposeOpDim::WH: // THIS NEEDS TO BE FIXED
+            if (((W * a.element_size()) % ROW_MAJOR_STICK_WIDTH != 0) || ((H * a.element_size()) % ROW_MAJOR_STICK_WIDTH != 0)) {
+                tiled_only = true;
+            }
+            else if (a.device()->arch() == tt::ARCH::GRAYSKULL) {
+                tiled_only = a.shape()[-2] > 256; // horrible hack because PCC on transpose HW row major is terrible on GS in this code path - kernel spits out garbage and has some demuxing for greater than this size that doesn't work
+            }
+            else if (a.device()->arch() == tt::ARCH::WORMHOLE_B0) {
+                tiled_only = !a.is_sharded() && (a.shape()[-2]*a.shape()[-1] >= 400000); // CB blows up on large sizes, hack until transpose_wh_rm is optimized
```

Comment:
Is this magic number here because we don't have a way to query available space for CB allocation? If there was a way to query usable space for CBs would this go away? If so, is this a number you picked based on standalone op traces? Is it realistic to expect it to work when the op is run within a larger model?